fix: small fixes to support boost multiprecision and rational

barendgehrels opened the pull request:
This PR is a continuation of  #901 which was first to be merged into the other branch, and it's pr declined, but later split off

barendgehrels commented on include/boost/geometry/algorithms/detail/overlay/sort_by_side.hpp:
```diff
@@ -316,7 +316,15 @@ public :
+            <
+                typename geometry::coordinate_type<Point>::type,
+                double
+            >::type;
```
Required for types as `std::int16_t` to avoid overflow
(therefore it's tested in `union_other_types.cpp` now too)

barendgehrels commented on include/boost/geometry/algorithms/detail/sections/sectionalize.hpp:
```diff
@@ -783,22 +783,24 @@ inline void enlarge_sections(Sections& sections, Strategy const&)
-        detail::buffer::buffer_box(it->bounding_box, 0.001, it->bounding_box);
+        using gt = decltype(section.bounding_box);
+        using ct = typename geometry::coordinate_type<gt>::type;
+        ct const tolerance = ct(1) / ct(1000);
```
Required for `Boost.Rational`

barendgehrels commented:
Note, I get errors in `algorithms_convex_hull_robust`, unrelated to this PR.
This might be related to the merging of `use_side_robust_predicate`

vissarion commented:
> Note, I get errors in `algorithms_convex_hull_robust`, unrelated to this PR.
> This might be related to the merging of `use_side_robust_predicate`

Thanks for noticing! https://github.com/boostorg/geometry/pull/908 should fix that.

vissarion reviewed (approved):
Thanks! It would be nice to have a list of all Boost multi-precision types that are supported. Are they all?

barendgehrels commented:
> Thanks! It would be nice to have a list of all Boost multi-precision types that are supported. Are they all?

Good point. Supported are the ones tested here. But there are more, I'll check.

barendgehrels commented:
> Thanks! It would be nice to have a list of all Boost multi-precision types that are supported. Are they all?

All types supported by Boost multi-precision are now included in the unit test. Those are (basically) all types but not types using other libraries (GMP, tom, ...) and not float128 (compiler dependent). Also finished the intended testcases, it was actually hard finding a test going wrong without rescaling (a good sign).

Also added some additional small `zero` constants for boost::rational<short> (requires somehow more than other types)

I squashed it, and will merge.
